src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java

```diff
@@ -4294,4 +4294,67 @@ public BufferedImage getMascotImage() {
     public List<Integer> getAllHeldItems() {
         return Gen4Constants.allHeldItems;
     }
+
+    @Override
+    public List<Integer> getSensibleHeldItemsFor(TrainerPokemon tp, boolean consumableOnly, List<Move> moves, Map<Integer, List<MoveLearnt>> movesets) {
+        List<Integer> items = new ArrayList<>();
+        items.addAll(Gen4Constants.generalPurposeConsumableItems);
+        if (!consumableOnly) {
+            items.addAll(Gen4Constants.generalPurposeItems);
+        }
+        int[] pokeMoves = RomFunctions.getMovesAtLevel(tp.pokemon.number, movesets, tp.level);
+        for (int moveIdx : pokeMoves) {
+            Move move = moves.get(moveIdx);
+            if (move == null) {
+                continue;
+            }
+            if (move.category == MoveCategory.PHYSICAL) {
+                items.add(Gen4Constants.liechiBerry);
+                if (!consumableOnly) {
+                    items.addAll(Gen4Constants.typeBoostingItems.get(move.type));
+                    items.add(Gen4Constants.choiceBand);
+                    items.add(Gen4Constants.muscleBand);
+                }
+            }
+            if (move.category == MoveCategory.SPECIAL) {
+                items.add(Gen4Constants.petayaBerry);
+                if (!consumableOnly) {
+                    items.addAll(Gen4Constants.typeBoostingItems.get(move.type));
+                    items.add(Gen4Constants.wiseGlasses);
+                    items.add(Gen4Constants.choiceSpecs);
+                }
+            }
+            if (!consumableOnly && Gen4Constants.moveBoostingItems.containsKey(moveIdx)) {
+                items.addAll(Gen4Constants.moveBoostingItems.get(moveIdx));
+            }
+        }
+        Map<Type, Effectiveness> byType = Effectiveness.against(tp.pokemon.primaryType, tp.pokemon.secondaryType, 4);
+        for(Map.Entry<Type, Effectiveness> entry : byType.entrySet()) {
+            Integer berry = Gen4Constants.weaknessReducingBerries.get(entry.getKey());
+            if (entry.getValue() == Effectiveness.DOUBLE) {
+                items.add(berry);
+            } else if (entry.getValue() == Effectiveness.QUADRUPLE) {
+                for (int i = 0; i < 8; i ++) {  // Increase the likelihood of protecting against 4x weak.
+                    items.add(berry);
+                }
+            }
+        }
+
+        if (!consumableOnly) {
+            if (tp.ability == Gen4Constants.guts) {
+                items.add(Gen4Constants.flameOrb);
+                items.add(Gen4Constants.toxicOrb);
+            }
+            if (tp.pokemon.primaryType == Type.POISON || tp.pokemon.secondaryType == Type.POISON) {
+                items.add(Gen4Constants.blackSludge);

```

Comment:
Wouldn't it make more sense for Black Sludge to be put in the "Type Boosting Items" map, instead of giving it its own case like this?

Reply:
typeBoosting item cares about the move type. Black sludge cares about the pokemon type.

Reply:
Right, my bad.